src/mpid/ch4/shm/posix/eager/iqueue/iqueue_recv.h

```diff
@@ -10,34 +10,39 @@
 #include "mpidu_genq.h"
 
 MPL_STATIC_INLINE_PREFIX int
-MPIDI_POSIX_eager_recv_begin(MPIDI_POSIX_eager_recv_transaction_t * transaction)
+MPIDI_POSIX_eager_recv_begin(int vsi, MPIDI_POSIX_eager_recv_transaction_t * transaction)
 {
     MPIDI_POSIX_eager_iqueue_transport_t *transport;
     MPIDI_POSIX_eager_iqueue_cell_t *cell = NULL;
     int ret = MPIDI_POSIX_NOK;
 
     MPIR_FUNC_ENTER;
 
-    /* Get the transport with the global variables */
-    transport = MPIDI_POSIX_eager_iqueue_get_transport();
-
-    MPIDU_genq_shmem_queue_dequeue(transport->cell_pool, transport->my_terminal, (void **) &cell);
-
-    if (cell) {
-        transaction->src_local_rank = cell->from;
-        transaction->payload = MPIDI_POSIX_EAGER_IQUEUE_CELL_PAYLOAD(cell);
-        transaction->payload_sz = cell->payload_size;
-
-        if (likely(cell->type == MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_HDR)) {
-            transaction->msg_hdr = &cell->am_header;
-        } else {
-            MPIR_Assert(cell->type == MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_DATA);
-            transaction->msg_hdr = NULL;
+    /* TODO: measure the latency overhead due to multiple vsi */
```

Comment:
We should look at single-thread performance in all areas before and after all the vci additions. We're adding some things (like modulos) to the critical path that might have quantifiable impact.

Reply:
The modulos are bypassed in single-thread build. Even with runtime single vci, the branch is there but modulo is skipped.

Reply:
To be clear, it will return constant `0` in `MPIDI_get_vci` (`ch4/src/ch4_vci.h`). Since that is inlined, the compiler should be able to optimize the modulo away.